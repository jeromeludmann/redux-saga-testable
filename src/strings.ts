const TAB_WIDTH = 2;
const MAX_DEPTH = 6;

type StringifierOptions = Partial<{
  tabWidth: number;
  maxDepth: number;
}>;

class Stringifier {
  private readonly tabWidth: string;
  private readonly maxDepth: number;
  private currentWidth: string;
  private currentDepth: number;

  constructor(opts: StringifierOptions) {
    this.tabWidth = new Array(opts.tabWidth ?? TAB_WIDTH).fill(' ').join('');
    this.maxDepth = opts.maxDepth ?? MAX_DEPTH;
    this.currentWidth = this.tabWidth;
    this.currentDepth = 0;
  }

  stringify(value: unknown): string {
    return `${this.tabWidth}${this.stringifyValue(value)}`;
  }

  private stringifyValue(value: any): string {
    switch (typeof value) {
      case 'undefined':
        return 'undefined';

      case 'string':
        return `"${value}"`;

      case 'function':
        return `\u0192 ${value.name || '<anonymous>'}`;

      case 'object':
        return this.stringifyObject(value);

      default:
        return value.toString();
    }
  }

  private stringifyObject(value: any): string {
    if (value === null) {
      return 'null';
    }

    if (Array.isArray(value)) {
      return this.stringifyArray(value);
    }

    if (value.constructor !== Object) {
      return value.toString();
    }

    const keys = [
      ...Object.keys(value),
      ...Object.getOwnPropertySymbols(value),
    ];

    return this.stringifyNested(
      () =>
        keys.map((key) => `${String(key)}: ${this.stringifyValue(value[key])}`),
      { wrapper: '{}', size: keys.length },
    );
  }

  private stringifyArray(value: unknown[]): string {
    return this.stringifyNested(
      () => value.map((value) => this.stringifyValue(value)),
      {
        wrapper: '[]',
        size: value.length,
      },
    );
  }

  private stringifyNested(
    stringifyFn: () => string[],
    opts: { wrapper: '{}' | '[]'; size: number },
  ): string {
    const [start, end] = opts.wrapper.split('');
    const wrap = (str: string): string => `${start}${str}${end}`;

    if (this.currentDepth === this.maxDepth) {
      return `${wrap(`\u2026`)} (${opts.size})`;
    }

    const prevWidth = this.currentWidth;

    this.currentWidth += this.tabWidth;
    ++this.currentDepth;

    const elements = stringifyFn();

    if (elements.length === 0) {
      return wrap('');
    }

    const nested = elements
      .map((str) => `${this.currentWidth}${String(str)}`)
      .join(',\n');

    this.currentWidth = prevWidth;

    return `${wrap(`\n${nested}\n${prevWidth}`)}`;
  }
}

export function stringify(
  value: unknown,
  { tabWidth, maxDepth }: StringifierOptions = {},
): string {
  return new Stringifier({ tabWidth, maxDepth }).stringify(value);
}
