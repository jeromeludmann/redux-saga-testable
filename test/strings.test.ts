/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { stringify } from 'redux-saga-testable/strings';

describe('stringify()', () => {
  test('undefined', () => {
    const undefinedValue = undefined;
    expect(stringify(undefinedValue)).toMatchSnapshot();
  });

  test('null', () => {
    const nullValue = null;
    expect(stringify(nullValue)).toMatchSnapshot();
  });

  test('number', () => {
    const number = 123;
    expect(stringify(number)).toMatchSnapshot();
  });

  test('string', () => {
    const string = 'string';
    expect(stringify(string)).toMatchSnapshot();
  });

  test('regexp', () => {
    const regexp = /^regexp$/;
    expect(stringify(regexp)).toMatchSnapshot();
  });

  test('function', () => {
    const fn = () => true;
    expect(stringify(fn)).toMatchSnapshot();
  });

  test('anonymous function', () => {
    expect(stringify(() => true)).toMatchSnapshot();
  });

  test('Map', () => {
    const map = new Map();
    expect(stringify(map)).toMatchSnapshot();
  });

  test('Set', () => {
    const set = new Set();
    expect(stringify(set)).toMatchSnapshot();
  });

  test('Symbol', () => {
    const symbol = { [Symbol('key')]: { symbol: Symbol('value') } };
    expect(stringify(symbol)).toMatchSnapshot();
  });

  test('array with custom spaces', () => {
    const array = ['value1', 'value2', 'value3'];

    expect(stringify(array, { tabWidth: 2 })).toMatchSnapshot();
    expect(stringify(array, { tabWidth: 4 })).toMatchSnapshot();
  });

  test('empty array', () => {
    const emptyArray: any[] = [];
    expect(stringify(emptyArray)).toMatchSnapshot();
  });

  test('array with depth level', () => {
    const deepArray = [[[[[['phew!']]]]]];

    expect(stringify(deepArray, { maxDepth: 3 })).toMatchSnapshot();
    expect(stringify(deepArray, { maxDepth: 6 })).toMatchSnapshot();
  });

  test('object with custom spaces', () => {
    const object = {
      key1: 'value1',
      key2: 'value2',
      key3: 'value3',
    };

    expect(stringify(object, { tabWidth: 2 })).toMatchSnapshot();
    expect(stringify(object, { tabWidth: 4 })).toMatchSnapshot();
  });

  test('empty object', () => {
    const emptyObject = {};
    expect(stringify(emptyObject)).toMatchSnapshot();
  });

  test('object with depth level', () => {
    const deepObject = {
      key1: {
        key2: {
          key3: {
            key4: {
              key5: {
                key6: 'phew!',
              },
            },
          },
        },
      },
    };

    expect(stringify(deepObject, { maxDepth: 3 })).toMatchSnapshot();
    expect(stringify(deepObject, { maxDepth: 6 })).toMatchSnapshot();
  });
});
