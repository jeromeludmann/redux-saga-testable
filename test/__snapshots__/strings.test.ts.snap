// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringify() Map 1`] = `"  [object Map]"`;

exports[`stringify() Symbol 1`] = `
"  {
    Symbol(key): {
      symbol: Symbol(value)
    }
  }"
`;

exports[`stringify() anonymous function 1`] = `"  ƒ <anonymous>"`;

exports[`stringify() array with custom spaces 1`] = `
"  [
    \\"value1\\",
    \\"value2\\",
    \\"value3\\"
  ]"
`;

exports[`stringify() array with custom spaces 2`] = `
"    [
        \\"value1\\",
        \\"value2\\",
        \\"value3\\"
    ]"
`;

exports[`stringify() array with depth level 1`] = `
"  [
    [
      [
        […] (1)
      ]
    ]
  ]"
`;

exports[`stringify() array with depth level 2`] = `
"  [
    [
      [
        [
          [
            [
              \\"phew!\\"
            ]
          ]
        ]
      ]
    ]
  ]"
`;

exports[`stringify() empty array 1`] = `"  []"`;

exports[`stringify() empty object 1`] = `"  {}"`;

exports[`stringify() function 1`] = `"  ƒ fn"`;

exports[`stringify() null 1`] = `"  null"`;

exports[`stringify() number 1`] = `"  123"`;

exports[`stringify() object with custom spaces 1`] = `
"  {
    key1: \\"value1\\",
    key2: \\"value2\\",
    key3: \\"value3\\"
  }"
`;

exports[`stringify() object with custom spaces 2`] = `
"    {
        key1: \\"value1\\",
        key2: \\"value2\\",
        key3: \\"value3\\"
    }"
`;

exports[`stringify() object with depth level 1`] = `
"  {
    key1: {
      key2: {
        key3: {…} (1)
      }
    }
  }"
`;

exports[`stringify() object with depth level 2`] = `
"  {
    key1: {
      key2: {
        key3: {
          key4: {
            key5: {
              key6: \\"phew!\\"
            }
          }
        }
      }
    }
  }"
`;

exports[`stringify() regexp 1`] = `"  /^regexp$/"`;

exports[`stringify() string 1`] = `"  \\"string\\""`;

exports[`stringify() undefined 1`] = `"  undefined"`;
